### ISSUE TITLE
Add profile system to specify unique cases for breakpoints, reading values, etc.

### ISSUE THREAD

**Issue by chipsenkbeil** [chipsenkbeil/scala-debugger#30]
One of the high-level features that I think is needed would be a debug profile. The idea behind this is that you have a low-level, pure JDI API and a higher-level set of rules on how to transform normal JDI operations into something better for Scala.

The low level API is untainted and handles breakpoints, method entry, method exit, exception, etc. with no special cases. Essentially, it acts as a Java debugger.

The higher-level set of rules is the profile, which provides details for specific gotchas caused by the code Scala generates. For instance, the profile could include logic to continuously step into and over different lines of code to find the next relevant piece of Scala when stepping through generated Scala classes (loops, collections, etc). Or, it might help ignore stepping onto the same line multiple times, which can happen when resuming a breakpoint on an assignment in Scala as it invokes a `{var}_$eq` method.

```scala
object Main extends App {
  val x = 3 // A breakpoint is triggered twice on this line due to x_$eq
  var y = 4 // The same for this line, due to y_$eq

  println(x + y)
}
```

The reason for providing a profile system versus hard-coding cases is due to differences between Scala versions (e.g. 2.10 and 2.11 have different breakpointable lines) and different ways of running Scala (a normal jar vs the REPL, which generates a series of nested classes with your code in the middle). Furthermore, providing some way to plug in custom handlers for these situations makes - to me - this library more extensible and future-proof. With Scala 2.12 adopting parts of JDK 8's new functionality (hence why it requires JDK 8), being able to inject a completely different set of rules on top of the low-level API seems like a must.

Currently, there is only one place in the debugger that has a hard-coded Scala conditional, which is [here](https://github.com/chipsenkbeil/scala-debugger/blob/7002ed6924e556b9fc95d300d5eb255ffa7ce9e8/scala-debugger-api/src/main/scala/org/senkbeil/debugger/wrappers/ValueWrapper.scala#L133). The value wrapper ignores any field with the name _MODULE$_ when retrieving values as to avoid recursion, although that might not be necessary by default (need to double check my logic for introducing that into the low level API).

This is something that I really want to get feedback from others. If this system is diverse enough, I think it will make maintaining specific cases much easier. But, getting input from others interested/involved in this space is definitely a must.

**Comment by chipsenkbeil** [comment]
So, things that need to be discussed for this:

1. How do people specify rules for the profile? Is it by implementing some common interface? Providing config files?
2. What kind of JDI events does a profile listen to? What can a profile modify?
3. What kinds of actions can a profile normally take? How low can it go? Can it just use filters to ignore certain events? (probably needs more control)
4. What should the default profile be? (a generic Scala one that covers most cases?)

**Comment by chipsenkbeil** [comment]
I've implemented the initial version to be available in v1.1.0. I'm still deciding on how to handle errors when performing logic like the following:

```scala
val s: ScalaVirtualMachine = _

// The onBreakpoint method could throw a variety of exceptions
// including the file or line not being available or the VM being disconnected
s.onBreakpoint("somefile.scala", 33)
  .map((e: Event) => e.location())
  .map((l: Location) => (l.sourcePath(), l.lineNumber()))
  .foreach((info: (String, Int)) => println(s"Reached breakpoint at ${info._1}:${info._2}"))
```

**Comment by chipsenkbeil** [comment]
Using Try[Pipeline[..., ...]] to handle errors while also provide an unsafe variation that will throw the exception caught by the Try.

```scala
val s: ScalaVirtualMachine = _

s.onUnsafeBreakpoint("somefile.scala", 33)
  .map((e: Event) => e.location())
  .map((l: Location) => (l.sourcePath(), l.lineNumber()))
  .foreach((info: (String, Int)) => println(s"Reached breakpoint at ${info._1}:${info._2}"))
```
